### PR TITLE
Work branch: Guard middleware

### PR DESCRIPTION
**Pull request by erastian**
- added guard middleware
- updated articles, auth, categories and users controllers for using guard middleware
- added "Publish" method to articles module
- Added "url" field in Articles model
- create and update DTOs was updated
- Updated Articles module with a lot of internal methods

TODO: Comments module still in another branch. Need to come back to this module after comments module merge in main branch. Article module needs a revision of the methods

### CODE REVIEW

File: src/articles/articles.controller.ts [hunk 1]
```diff
@@ -26,15 +30,35 @@ export class ArticlesController extends BaseController implements IArticlesContr
...
+					new GuardMiddleware([Role.ADMIN, Role.MANAGER]),
+					new ValidateMiddleware(ArticleUpdateDto),
+				],
+			},
+			{
+				path: '/:id/publish',
+				method: 'post',
+				func: this.publishArticle,
+				middlewares: [
+					new GuardMiddleware([Role.ADMIN, Role.MANAGER]),
```
Comment (mdziarkach) [inline src/articles/articles.controller.ts]: only admin

File: src/articles/articles.controller.ts [hunk 2]
```diff
@@ -78,6 +102,19 @@ export class ArticlesController extends BaseController implements IArticlesContr
...
+
+			this.ok(res, result);
+		} catch (e) {
+			return next(e);
+		}
+	}
+
 	async deleteArticle({ params }: Request, res: Response, next: NextFunction): Promise<void> {
 		try {
 			const id = params.id;
```
Comment (mdziarkach) [inline src/articles/articles.controller.ts]: if manager didn't create this article then he can't delete it

File: src/articles/articles.controller.ts [hunk 3]
```diff
@@ -26,15 +30,35 @@ export class ArticlesController extends BaseController implements IArticlesContr
...
+					new ValidateMiddleware(ArticleCreateDto),
+				],
 			},
 			{
 				path: '/:id',
 				method: 'patch',
 				func: this.updateArticle,
-				middlewares: [new ValidateMiddleware(ArticleUpdateDto)],
+				middlewares: [
+					new GuardMiddleware([Role.ADMIN, Role.MANAGER]),
```
Comment (mdziarkach) [inline src/articles/articles.controller.ts]: manager can update only own article

File: src/common/guard.middleware.ts [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+import { HTTPError } from '../services/errors/http-error.class';
+
+import { IMiddleware } from './middleware.interface';
+
+export class GuardMiddleware implements IMiddleware {
+	constructor(private roles: User['role'][] = []) {}
+
+	execute(req: Request, res: Response, next: NextFunction): void {
+		if (!req.user) {
+			return next(new HTTPError(StatusCodes.FORBIDDEN, 'User not authorized'));
```
Comment (mdziarkach) [inline src/common/guard.middleware.ts:13]: let's return 401 status code here

File: src/articles/articles.controller.ts (hunk 1 above)
Comment (erastian) [reply src/articles/articles.controller.ts]: Done

File: src/common/guard.middleware.ts (hunk 4 above)
Comment (erastian) [reply src/common/guard.middleware.ts:13]: Done

File: src/articles/articles.controller.ts (hunk 2 above)
Comment (erastian) [reply src/articles/articles.controller.ts]: Done

File: src/articles/articles.controller.ts (hunk 3 above)
Comment (erastian) [reply src/articles/articles.controller.ts]: Done

**Pull request merged**